core/templates/pages/exploration-editor-page/translation-tab/audio-translation-bar/audio-translation-bar.directive.spec.ts

```diff
@@ -0,0 +1,747 @@
+
+    isTranslatableSpy = spyOn(editabilityService, 'isTranslatable');
+    isTranslatableSpy.and.returnValue(false);
+    spyOn(translationLanguageService, 'getActiveLanguageCode').and
+      .returnValue('en');
+    spyOn(translationTabActiveContentIdService, 'getActiveContentId').and
+      .returnValue('content');
+    spyOn(stateEditorService, 'getActiveStateName').and.returnValue(
+      stateName);
+    // This method is being mocked because this spec only handles with
+    // recordedvoiceovers and not all the exploration.
+    spyOn(explorationStatesService, 'saveRecordedVoiceovers').and
+      .callFake(function() {});
+
+    stateRecordedVoiceoversService.init(stateName,
+      recordedVoiceoversObjectFactory.createFromBackendDict({
+        voiceovers_mapping: {
+          content: {
+            en: {
+              duration_secs: 0,
+              filename: '',
+              file_size_bytes: 0,
+              needs_update: false
+            }
+          }
+        }
+      }));
+
+    var directive = $injector.get('audioTranslationBarDirective')[0];
+    ctrl = $injector.instantiate(directive.controller, {
+      $scope: $scope,
+      AlertsService: alertsService,
+      StateRecordedVoiceoversService: stateRecordedVoiceoversService
+    });
+    ctrl.$onInit();
+    $scope.getVoiceoverRecorder();
+  }));
+
+  afterEach(function() {
+    $rootScope.$broadcast('$destroy');
+  });
+
+  it('should evaluate $scope properties after audio bar initialization',
+    function() {
+      expect($scope.languageCode).toBe('en');
+      expect($scope.contentId).toBe('content');
+    });
+
+  it('should not check and start recording when user deny access',
+    function() {
+      spyOn(voiceoverRecordingService, 'status').and.returnValue({
+        isAvailable: true
+      });
+      spyOn(voiceoverRecordingService, 'startRecording').and.returnValue(
+        $q.reject());
+
+      $scope.checkAndStartRecording();
+      $scope.$apply();
+
+      expect($scope.unsupportedBrowser).toBe(false);
+      expect($scope.recordingPermissionDenied).toBe(true);
+      expect($scope.cannotRecord).toBe(true);
+    });
+
+  it('should not check and start recording when voiceover recorder is' +
+    ' not available', function() {
+    spyOn(voiceoverRecordingService, 'status').and.returnValue({
+      isAvailable: false
+    });
+    $scope.checkAndStartRecording();
+
+    expect($scope.unsupportedBrowser).toBe(true);
+    expect($scope.cannotRecord).toBe(true);
+  });
+
+  it('should stop recording when reaching recording time limit', function() {
+    spyOn(voiceoverRecordingService, 'status').and.returnValue({
+      isAvailable: true
+    });
+    spyOn(voiceoverRecordingService, 'startRecording').and.returnValue(
+      $q.resolve());
+    spyOn($scope.voiceoverRecorder, 'getMp3Data').and.returnValue(
+      $q.resolve([]));
+
+    $scope.checkAndStartRecording();
+    $scope.$apply();
+
+    $scope.elapsedTime = 298;
+    $interval.flush(1000);
```

Comment:
Are we sure this isn't flaky.

Reply:
The `$interval` is more stable than `$timeout`, its usage is not so flicky because it uses a [different queue](https://docs.angularjs.org/api/ngMock/service/$flushPendingTasks). Also, the documentation says that the best way to test `$interval` is by using `$interval.flush`.